core/src/main/scala/org/apache/spark/api/python/StreamingPythonRunner.scala

```diff
@@ -76,47 +78,60 @@ private[spark] class StreamingPythonRunner(
     pythonWorker = Some(worker)
     pythonWorkerFactory = Some(workerFactory)
 
-    logInfo(log"[session: ${MDC(SESSION_ID, sessionId)}] Python worker created")
-
-    val stream = new BufferedOutputStream(
-      pythonWorker.get.channel.socket().getOutputStream, bufferSize)
+    val socket = pythonWorker.get.channel.socket()
+    val stream = new BufferedOutputStream(socket.getOutputStream, bufferSize)
+    val dataIn = new DataInputStream(new BufferedInputStream(socket.getInputStream, bufferSize))
     val dataOut = new DataOutputStream(stream)
 
-    logInfo(log"[session: ${MDC(SESSION_ID, sessionId)}] Sending necessary information to the " +
-      log"Python worker")
-
-    PythonWorkerUtils.writePythonVersion(pythonVer, dataOut)
+    val originalTimeout = socket.getSoTimeout()
+    // Set timeout to 5 minute during initialization config transmission
+    socket.setSoTimeout(5 * 60 * 1000)
 
-    // Send sessionId
-    if (!sessionId.isEmpty) {
-      PythonRDD.writeUTF(sessionId, dataOut)
-    }
+    val resFromPython = try {
+      PythonWorkerUtils.writePythonVersion(pythonVer, dataOut)
 
-    // Send the user function to python process
-    PythonWorkerUtils.writePythonFunction(func, dataOut)
-    dataOut.flush()
+      // Send sessionId
+      if (!sessionId.isEmpty) {
+        PythonRDD.writeUTF(sessionId, dataOut)
+      }
 
-    logInfo(log"[session: ${MDC(SESSION_ID, sessionId)}] Reading initialization response from " +
-      log"Python runner.")
+      // Send the user function to python process
+      PythonWorkerUtils.writePythonFunction(func, dataOut)
+      dataOut.flush()
 
-    val dataIn = new DataInputStream(
-      new BufferedInputStream(pythonWorker.get.channel.socket().getInputStream, bufferSize))
+      logInfo(log"[session: ${MDC(SESSION_ID, sessionId)}] Reading initialization response from " +
+        log"Python runner.")
+      dataIn.readInt()
+    } catch {
+      case e: java.net.SocketTimeoutException =>
+        throw new StreamingPythonRunnerInitializationTimeoutException(e.getMessage)
+      case e: Exception =>
+        throw new StreamingPythonRunnerInitializationCommunicationException(e.getMessage)
+    } finally {
+      // Set timeout back to the original timeout
+      // Should be infinity by default
+      socket.setSoTimeout(originalTimeout)
```

Comment:
I'm not familiar with Java's socket interface but is it safe to set it after some timeout against the socket without exception? If you've checked and it never throws exceptions I'm OK with this.